modules/FindImGui.cmake

```diff
@@ -0,0 +1,107 @@
+#
+#  ImGui_INCLUDE_DIR          - Include dir
+#
+
+#
+#   This file is part of Magnum.
+#
+#   Copyright © 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018
+#             Vladimír Vondruš <mosra@centrum.cz>
+#   Copyright © 2018 Jonathan Hale <squareys@googlemail.com>
+#
+#   Permission is hereby granted, free of charge, to any person obtaining a
+#   copy of this software and associated documentation files (the "Software"),
+#   to deal in the Software without restriction, including without limitation
+#   the rights to use, copy, modify, merge, publish, distribute, sublicense,
+#   and/or sell copies of the Software, and to permit persons to whom the
+#   Software is furnished to do so, subject to the following conditions:
+#
+#   The above copyright notice and this permission notice shall be included
+#   in all copies or substantial portions of the Software.
+#
+#   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+#   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+#   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+#   DEALINGS IN THE SOFTWARE.
+#
+
+find_package(imgui CONFIG)
+if(imgui_FOUND)
+    if(NOT TARGET ImGui::ImGui)
+        add_library(ImGui::ImGui INTERFACE IMPORTED)
+        set_property(TARGET ImGui::ImGui APPEND PROPERTY
+            INTERFACE_LINK_LIBRARIES imgui::imgui)
+
+        get_target_property(ImGui_INCLUDE_DIR imgui::imgui
+            INTERFACE_INCLUDE_DIRECTORIES)
+
+        add_library(ImGui::Sources INTERFACE IMPORTED)
+        set_property(TARGET ImGui::Sources APPEND PROPERTY
+            INTERFACE_LINK_LIBRARIES ImGui::ImGui)
+    endif()
+else()
+    find_path(ImGui_INCLUDE_DIR NAMES imgui.h HINTS ${IMGUI_DIR})
+
+    if(NOT TARGET ImGui::ImGui)
+        add_library(ImGui::ImGui INTERFACE IMPORTED)
+        set_property(TARGET ImGui::ImGui APPEND PROPERTY
+            INTERFACE_INCLUDE_DIRECTORIES "${ImGui_INCLUDE_DIR}")
+    endif()
+endif()
+
+# Find components
+foreach(_component IN LISTS ImGui_FIND_COMPONENTS)
+
+    if(_component STREQUAL "Sources")
+        find_file(ImGui_SOURCE NAMES imgui.cpp HINTS ${IMGUI_DIR})
+        set(ImGui_SOURCES ${ImGui_SOURCE})
+
+        if(NOT ImGui_SOURCE)
+            set(ImGui_Sources_FOUND TRUE)
+        else()
+            set(ImGui_Sources_FOUND FALSE)
+        endif()
+
+        foreach(_file widgets draw demo)
+            find_file(ImGui_${_file}_SOURCE NAMES imgui_${_file}.cpp
+                HINTS ${IMGUI_DIR})
+            list(APPEND ImGui_SOURCES ${ImGui_${_file}_SOURCE})
+            if(NOT ImGui_${_file}_SOURCE)
+                set(ImGui_Sources_FOUND FALSE)
+            endif()
+        endforeach()
```

Comment:
Compiling ImGui sources as part of the client app will cause lots of warnings emitted on GCC, this could be (I hope) a solution:

```cmake
# Clang-specific flags
if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR (CMAKE_CXX_COMPILER_ID MATCHES "(Apple)?Clang" AND NOT CMAKE_CXX_SIMULATE_ID STREQUAL "MSVC") OR CORRADE_TARGET_EMSCRIPTEN)
    set_property(FILE ${ImGui_${_file}_SOURCE} APPEND PROPERTY COMPILE_FLAGS 
        "-Wno-zero-as-nullptr-constant")
endif()

# GCC-specific flags
if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
    set_property(FILE ${ImGui_${_file}_SOURCE} APPEND PROPERTY COMPILE_FLAGS 
        "-Wno-double-promotion")
endif()
```

Reply:
It appears that `FILE` is not a valid scope for `set_property`. Full error:
```
CMake Error at modules/FindImGui.cmake:95 (set_property):
  set_property given invalid scope FILE.  Valid scopes are GLOBAL, DIRECTORY,
  TARGET, SOURCE, TEST, CACHE, INSTALL.
Call Stack (most recent call first):
  src/Magnum/ImGuiIntegration/CMakeLists.txt:30 (find_package)
```

Reply:
It was `set_property(SOURCE ... APPEND_STRING ...`, works, but still spews out a lot of warnings...

Reply:
Ah, wait, my bad

Reply:
I had to use spaces in front of the flag now with `APPEND_STRING`, obviously. The nullptr flag is spelled `null-pointer` and apart from that I also added `-Wno-old-style-cast` on GNU and clang and now it's silent! 👏 👏